design/src/rfcs/rfc0003_presigning_api.md

```diff
@@ -0,0 +1,208 @@
+RFC: API for Presigned URLs
+============================
+
+> Status: RFC
+
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+Several AWS services allow for presigned requests in URL form, which is described well by
+[S3's documentation on authenticating requests using query parameters](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html).
+
+This doc establishes the customer-facing API for creating these presigned URLs and how they will
+be implemented in a generic fashion in the SDK codegen.
+
+Terminology
+-----------
+
+To differentiate between the clients that are present in the generated SDK today, the following
+terms will be used throughout this doc:
+
+- **Smithy Client**: A `smithy_client::Client<C, M, R>` struct that is responsible for gluing together
+  the connector, middleware, and retry policy. This is not generated and lives in the `smithy-client` crate.
+- **Fluent Client**: A code-generated `Client<C, M, R>` that has methods for each service operation on it.
+  A fluent builder is generated alongside it to make construction easier.
+
+Presigned URL config
+--------------------
+
+Today, presigned URLs take an expiration time that's not part of the service API.
+The SDK will make this configurable as a separate struct so that there's no chance of name collisions, and so
+that additional fields can be added in the future. Fields added later will require defaulting for
+backwards compatibility.
+
+Customers should also be able to set a start time on the presigned URL's expiration so that they can
+generate URLs that become active in the future. An optional `start_time` option will be available and
+default to `SystemTime::now()`.
+
+Construction `PresigningConfig` can be done with a builder, but a `PresigningConfig::expires_in`
+convenience function will be provided to bypass the builder for the most frequent use-case.
+
+```rust
+#[non_exhaustive]
+#[derive(Debug, Clone)]
+pub struct PresigningConfig {
+    start_time: SystemTime,
+    expires_in: Duration,
+}
+
+#[non_exhaustive]
+#[derive(Debug)]
+pub struct Builder {
+    start_time: Option<SystemTime>,
+    expires_in: Option<Duration>,
+}
+
+impl Builder {
+    pub fn start_time(self, start_time: SystemTime) -> Self { ... }
+    pub fn set_start_time(self, start_time: Option<SystemTime>) -> Self { ... }
```

Comment:
nit: typically I have the setters take `&mut Self` which can make them a little more ergonomic to use.